### ISSUE TITLE
Implement better unrunnable schedule detection

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#16068]
The present unrunnable schedule detection uses an algorithm to find circular dependencies in the graph of module dependencies (where dependencies come both as data dependencies and order dependencies from Paths/EndPaths). The problem is the algorithm only finds a 'base set' of cycles and not all possible cycles. Therefore it is possible (and has actually happened in actual configurations) for the 'base cycle' to be a false positive (i.e. actually runnable based on other criteria) but a cycle which includes that 'base cycle' to itself be unrunnable.

Therefore we need to extend the algorithm to keep a list of the found 'base cycles' and then if the search algorithm encounters a node which is part of the base cycle, to evaluate if the extended cycle is unrunnable.

**Comment by Dr15Jones** [comment]
assign core

**Comment by Dr15Jones** [comment]
An example of a unrunnable schedule that we miss is
```
path p1: A + B + C +<PathEndedP1>
path p2: E + <PathEndedP2>
endpath e1: <EndPathBeganE1>+ A+D
```
Where we have the data dependencies
```
C depends on A
E depends on D
TriggerResults depends on <PathEndedP1> and <PathEndedP2>
<EndPathBeganE1> depends on TriggerResults
```

Doing a depth first search of the graph starting from `TriggerResults` (TR) gives
```
TR, PEP1
    PEP1, C
          C, A
             A, EPBE1
                EPBE1, TR <<< back_edge
           C, B
              B, A
TR, PEP2
    PEP2, E
          E, D
             D, A
```
The back edge `EPBE1, TR` is found to be a false postive since `A` is always run earlier on the path contains `C`.
The cycle of `TR->PEP2->E->D->A->BEPE1->TR` is not seen because the search does not go over verticies it has already seen (in this case `A`).

**Comment by Dr15Jones** [comment]
Fixed with #17102

**Comment by Dr15Jones** [comment]
+1
